### PR TITLE
Breakpoints' model

### PR DESCRIPTION
**Pull request by digeff**
Breakpoints' model for v2. When the client specifies how to set a breakpoint, we create a Breakpoint Recipie to store/model those instructions. These classes have methods to map them to instructions on a Script, or a URL, or URL Regexp to set the breakpoint in CDTP.

After we send the instructions to CDTP, the instructions might match a file, so we'll get the actual breakpoint that was set, and it's actual location.

We have different behaviors for breakpoints (breakpoints, logpoints, conditional breakpoints, hit count breakpoints). We model them by each breakpoint being related to a IBPActionWhenHit which stores the behavior that the breakpoint has when hit.

We use IBPRecipieStatus to model whether a breakpoint recipie matched a script, and has any actual locations associated with it (whether the breakpoint is bound or not).

We use BPRecipies because the setBreakpoints() API from the client deals with several breakpoints at a time, so it makes sense to handle all the breakpoints as a single object sometimes.

### CODE REVIEW

File: src/chrome/internal/breakpoints/bpActionWhenHit.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    public logMessage(logMessage: LogMessage): boolean {
+        return this.areSameClass(this._left, logMessage)
+            && this._left.expressionToLog === logMessage.expressionToLog;
+    }
+
+    private areSameClass<T extends IBPActionWhenHit>(left: IBPActionWhenHit, right: T): left is T {
+        return left.constructor === right.constructor;
+    }
+
+    constructor(private readonly _left: IBPActionWhenHit) { }
```
Comment (roblourens) [inline src/chrome/internal/breakpoints/bpActionWhenHit.ts]: constructor to top

File: src/chrome/internal/breakpoints/bpActionWhenHit.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+abstract class BaseBPActionWhenHit {
+    public abstract accept<T>(visitor: IBPActionWhenHitVisitor<T>): T;
+
+    public isEquivalentTo(bpActionWhenHit: IBPActionWhenHit): boolean {
+        return bpActionWhenHit.accept(new BPActionWhenHitIsEquivalentVisitor(this));
+    }
+}
+
+class BPActionWhenHitIsEquivalentVisitor implements IBPActionWhenHitVisitor<boolean> {
```
Comment (roblourens) [inline src/chrome/internal/breakpoints/bpActionWhenHit.ts]: I don't quite understand what this class is doing

File: src/chrome/internal/breakpoints/bpActionWhenHit.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+    private areSameClass<T extends IBPActionWhenHit>(left: IBPActionWhenHit, right: T): left is T {
+        return left.constructor === right.constructor;
+    }
+
+    constructor(private readonly _left: IBPActionWhenHit) { }
+}
+
+export class AlwaysPause extends BaseBPActionWhenHit {
+    public accept<T>(visitor: IBPActionWhenHitVisitor<T>): T {
```
Comment (roblourens) [inline src/chrome/internal/breakpoints/bpActionWhenHit.ts]: When is `accept` called? I don't understand how these are used.

File: src/chrome/internal/breakpoints/baseMappedBPRecipie.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+import { BaseBPRecipie, IBPRecipie, BPInScriptSupportedHitActions } from './bpRecipie';
+import { BPRecipieInSource } from './bpRecipieInSource';
+import { ILoadedSource } from '../sources/loadedSource';
+import { IScript } from '../scripts/script';
+import { createURLRegexp, URLRegexp } from '../locations/subtypes';
+import { utils } from '../../..';
+import { IURL } from '../sources/resourceIdentifier';
+import { CDTPScriptUrl } from '../sources/resourceIdentifierSubtypes';
+
+export interface IMappedBPRecipie<TResource extends ScriptOrSourceOrURLOrURLRegexp, TBPActionWhenHit extends IBPActionWhenHit>
```
[the remark this answers is not in the document]
Comment (digeff) [reply src/chrome/internal/breakpoints/baseMappedBPRecipie.ts]: Changed

File: src/chrome/internal/breakpoints/bpRecipieStatus.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        return `${this.recipie} is unbinded because ${this.statusDescription}`;
+    }
+
+    constructor(
+        public readonly recipie: IBPRecipie<ISource>,
+        public readonly statusDescription: string) {
+    }
+}
+
+export class BPRecipieIsBinded implements IBPRecipieStatus {
```
[the remark this answers is not in the document]
Comment (digeff) [reply src/chrome/internal/breakpoints/bpRecipieStatus.ts]: Changed

File: src/chrome/internal/breakpoints/bpActionWhenHit.ts (hunk 1 above)
Comment (digeff) [reply src/chrome/internal/breakpoints/bpActionWhenHit.ts]: Changed

File: src/chrome/internal/breakpoints/bpActionWhenHit.ts (hunk 2 above)
Comment (digeff) [reply src/chrome/internal/breakpoints/bpActionWhenHit.ts]: This is not necessary any more. Replacing with a simpler version.

File: src/chrome/internal/breakpoints/bpActionWhenHit.ts (hunk 3 above)
Comment (digeff) [reply src/chrome/internal/breakpoints/bpActionWhenHit.ts]: This is not necessary any more. Replacing with a simpler version.

File: src/chrome/internal/breakpoints/bpRecipes.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        });
+    }
+
+    public toString(): string {
+        return printArray(`BPs @ ${this.source}`, this.breakpoints);
+    }
+}
+
+export class BPRecipesInSource extends BaseBPRecipes<ISource> {
+    public tryResolving<R>(ifSuccesfulDo: (bpsInLoadedSource: BPRecipesInLoadedSource) => R, ifFaileDo: () => R): R {
```
Comment (mrcrane) [inline src/chrome/internal/breakpoints/bpRecipes.ts]: Should be successfulAction and failedAction

File: src/chrome/internal/breakpoints/breakpoint.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+abstract class BaseBreakpoint<TResource extends BPPossibleResources> implements IBreakpoint<TResource> {
+    public abstract get recipe(): IBPRecipe<TResource>;
+    public abstract get actualLocation(): ActualLocation<TResource>;
+
+    public toString(): string {
+        return `${this.recipe} actual location is ${this.actualLocation}`;
+    }
+}
+
+type MappeableBPPossibleResources = IScript | IResourceIdentifier<CDTPScriptUrl> | URLRegexp;
```
Comment (mrcrane) [inline src/chrome/internal/breakpoints/breakpoint.ts]: Mappable

File: src/chrome/internal/breakpoints/bpRecipeInSource.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+    /**
+     * Hit breakpoints are implemented by setting an always break breakpoint, and then auto-resuming until the hit condition is true.
+     * We use this method to create the always break breakpoint for a hit count breakpoint
+     */
+    public withAlwaysBreakAction(): BPRecipeInSource<AlwaysPause> {
+        return new BPRecipeInSource<AlwaysPause>(this.location, new AlwaysPause());
+    }
+
+    public tryResolvingSource<R>(succesfulAction: (breakpointInLoadedSource: BPRecipeInLoadedSource<TBPActionWhenHit>) => R,
```
[the remark this answers is not in the document]
Comment (digeff) [reply src/chrome/internal/breakpoints/bpRecipeInSource.ts]: Changed

File: src/chrome/internal/breakpoints/breakpoint.ts (hunk 7 above)
Comment (digeff) [reply src/chrome/internal/breakpoints/breakpoint.ts]: Changed

**Pull request merged**